### PR TITLE
Add backward and optimizer operator dependency pass.

### PR DESCRIPTION
**Pull request by gongweibao**
Add dependency between optimizer operators by the order of grads be allreduced. And it can speed up bert 7% on four v100 machines.

### CODE REVIEW

File: paddle/fluid/framework/details/CMakeLists.txt [hunk 1]
```diff
@@ -93,6 +93,6 @@ cc_library(build_strategy SRCS build_strategy.cc DEPS
         fuse_elewise_add_act_pass multi_batch_merge_pass 
         fuse_relu_depthwise_conv_pass
         memory_optimize_pass lock_free_optimize_pass
-        alloc_continuous_space_for_grad_pass fuse_all_reduce_op_pass
+        alloc_continuous_space_for_grad_pass fuse_all_reduce_op_pass backward_op_deps_pass
```
Comment (chengduoZH) [inline paddle/fluid/framework/details/CMakeLists.txt]: This name is confusing.

File: paddle/fluid/framework/ir/alloc_continuous_space_for_grad_pass.cc [hunk 2]
```diff
@@ -207,13 +206,29 @@ class AllocContinuousSpaceForGradPass : public ir::Pass {
     }
 
     VLOG(10) << "SetGroupAccordingToLayers: ";
+    PrintGroupInfo(var_nodes, group_grads_params);
```
Comment (chengduoZH) [inline paddle/fluid/framework/ir/alloc_continuous_space_for_grad_pass.cc]:
```
if(VLOG_IS_ON(10)){
  PrintGroupInfo(var_nodes, group_grads_params);
}
```

File: paddle/fluid/framework/ir/alloc_continuous_space_for_grad_pass.cc [hunk 3]
```diff
@@ -258,60 +279,8 @@ class AllocContinuousSpaceForGradPass : public ir::Pass {
...
-                                group_size);
-    for (size_t i = 0; i < group_grads_params->size(); ++i) {
-      VLOG(10) << "group " << i;
-      std::stringstream out;
-      for (auto &p_g : group_grads_params->at(i)) {
-        out << "(" << p_g.second << ", " << p_g.first << "), ";
-      }
-      VLOG(10) << out.str();
-    }
+    PrintGroupInfo(var_nodes, group_grads_params);
```
Comment (chengduoZH) [inline paddle/fluid/framework/ir/alloc_continuous_space_for_grad_pass.cc]: Same above.

File: paddle/fluid/framework/ir/multi_devices_graph_pass/backward_op_deps_pass.cc [hunk 4]
```diff
@@ -0,0 +1,221 @@
+// Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
```
Comment (chengduoZH) [inline paddle/fluid/framework/ir/multi_devices_graph_pass/backward_op_deps_pass.cc]: 2018 -> 2019

File: paddle/fluid/operators/alloc_continuous_space_op.cc [hunk 5]
```diff
@@ -105,10 +108,10 @@ class AllocContinuousSpaceKernel : public framework::OpKernel<T> {
           .Resize(dim);
       len = Alignment(len * size_of_dtype, context.GetPlace()) / size_of_dtype;
       offset += len;
-      VLOG(10) << "alloc_space_for_vars: output(" << out_var_names[i]
-               << ") ,dim:(" << dim << ")"
-               << " Address: " << out_tensors[i]->data<void>();
+      ss << "output(" << out_var_names[i] << ")  dim:(" << dim << ")"
+         << " address: " << out_tensors[i]->data<void>() << ", ";
     }
+    VLOG(10) << ss.str();
```
Comment (chengduoZH) [inline paddle/fluid/operators/alloc_continuous_space_op.cc:114]: Why change the above code?
If VLOG is not open, the debug info doesn't need to print.

File: paddle/fluid/framework/ir/alloc_continuous_space_for_grad_pass.cc [hunk 6]
```diff
@@ -23,15 +23,16 @@
 #include "paddle/fluid/framework/ir/graph_helper.h"
 #include "paddle/fluid/framework/op_registry.h"
 
-DEFINE_uint64(fuse_parameter_memory_size, 0,  // 0 KB
+DEFINE_uint64(fuse_parameter_memory_size, 0,  // Bytes
```
Comment (chengduoZH) [inline paddle/fluid/framework/ir/alloc_continuous_space_for_grad_pass.cc:26]: I think that maybe KB is better.

File: paddle/fluid/framework/ir/multi_devices_graph_pass/backward_op_deps_pass.cc [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+  }
+
+  void ApplyImpl(ir::Graph* graph) const override {
+    // NOTE: The operator nodes should be in topology order.
+    std::vector<details::OpHandleBase*> backward_op_handles;
+    std::vector<details::OpHandleBase*> all_opt_handles;
+    details::ParamsAndGrads params_grads;
+    std::vector<ir::Node*> topo_nodes = ir::TopologySortOperations(*graph);
+    for (auto& node : topo_nodes) {
+      if (!node->Op()) continue;
```
Comment (chengduoZH) [inline paddle/fluid/framework/ir/multi_devices_graph_pass/backward_op_deps_pass.cc]: Why the `node->Op()` maybe nullptr?

File: paddle/fluid/operators/alloc_continuous_space_op.cc (hunk 5 above)
Comment (gongweibao) [reply paddle/fluid/operators/alloc_continuous_space_op.cc:114]: It doesn't change the log level but the content. All the grouped vars will be printed in one line. It's more convenient

File: paddle/fluid/framework/details/CMakeLists.txt (hunk 1 above)
Comment (gongweibao) [reply paddle/fluid/framework/details/CMakeLists.txt]: Thanks! Done!

File: paddle/fluid/framework/ir/alloc_continuous_space_for_grad_pass.cc (hunk 2 above)
Comment (gongweibao) [reply paddle/fluid/framework/ir/alloc_continuous_space_for_grad_pass.cc]: Done.

File: paddle/fluid/framework/ir/alloc_continuous_space_for_grad_pass.cc (hunk 3 above)
Comment (gongweibao) [reply paddle/fluid/framework/ir/alloc_continuous_space_for_grad_pass.cc]: Done.

File: paddle/fluid/framework/ir/multi_devices_graph_pass/backward_op_deps_pass.cc (hunk 4 above)
Comment (gongweibao) [reply paddle/fluid/framework/ir/multi_devices_graph_pass/backward_op_deps_pass.cc]: Done.

File: paddle/fluid/framework/ir/alloc_continuous_space_for_grad_pass.cc (hunk 6 above)
Comment (gongweibao) [reply paddle/fluid/framework/ir/alloc_continuous_space_for_grad_pass.cc:26]: KB is not enough now. We use MB commonly.
But MB is more coarse-grained.

File: paddle/fluid/framework/ir/multi_devices_graph_pass/backward_optimizer_op_deps_pass.cc [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+    // NOTE: The operator nodes should be in topology order.
+    std::vector<details::OpHandleBase*> backward_op_handles;
+    std::vector<details::OpHandleBase*> all_opt_handles;
+    details::ParamsAndGrads params_grads;
+    std::vector<ir::Node*> topo_nodes = ir::TopologySortOperations(*graph);
+    for (auto& node : topo_nodes) {
+      if (!node->Op()) continue;
+
+      GetBackWardOpHandles(node, &backward_op_handles, &params_grads);
+      GetOptHandles(node, &all_opt_handles);
```
Comment (chengduoZH) [inline paddle/fluid/framework/ir/multi_devices_graph_pass/backward_optimizer_op_deps_pass.cc]: What's the meaning of `GetOptHandles` and `GetBackWardOpHandles `?

File: paddle/fluid/framework/ir/multi_devices_graph_pass/backward_optimizer_op_deps_pass.cc [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+    VLOG(10) << "backward_op_handles size:" << backward_op_handles.size()
+             << ", opt_handles size:" << all_opt_handles.size();
+
+    if (backward_op_handles.size() <= 1 || all_opt_handles.size() <= 1) {
+      VLOG(10) << "need not backward_op_deps_pass";
+      return;
+    }
+
+    std::vector<details::OpHandleBase*> opt_handles;
+    GetHeadOptHandles(all_opt_handles, &opt_handles, params_grads);
```
Comment (chengduoZH) [inline paddle/fluid/framework/ir/multi_devices_graph_pass/backward_optimizer_op_deps_pass.cc]: What's the meaning of `GetHeadOptHandles `?
The name is confusing.

File: paddle/fluid/framework/ir/multi_devices_graph_pass/backward_op_deps_pass.cc (hunk 7 above)
Comment (gongweibao) [reply paddle/fluid/framework/ir/multi_devices_graph_pass/backward_op_deps_pass.cc]: Node has no opdesc.

File: paddle/fluid/framework/ir/multi_devices_graph_pass/backward_optimizer_op_deps_pass.cc (hunk 8 above)
Comment (gongweibao) [reply paddle/fluid/framework/ir/multi_devices_graph_pass/backward_optimizer_op_deps_pass.cc]: GetBackWardOpHandles=> get operator handles which opdesc's role is OpRole::kBackward.
GetOptHandles => get operator handles which opdesc's role is OpRole::kOptimize.

File: paddle/fluid/framework/ir/multi_devices_graph_pass/backward_optimizer_op_deps_pass.cc (hunk 9 above)
Comment (gongweibao) [reply paddle/fluid/framework/ir/multi_devices_graph_pass/backward_optimizer_op_deps_pass.cc]: Get the roots of the optimizer handles' graphs.

Comment (gongweibao) [reply paddle/fluid/framework/ir/multi_devices_graph_pass/backward_optimizer_op_deps_pass.cc]: Done.

File: paddle/fluid/framework/ir/multi_devices_graph_pass/backward_optimizer_op_deps_pass.cc (hunk 8 above)
Comment (gongweibao) [reply paddle/fluid/framework/ir/multi_devices_graph_pass/backward_optimizer_op_deps_pass.cc]: Done.

File: paddle/fluid/framework/ir/multi_devices_graph_pass/backward_optimizer_op_deps_pass.cc [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+      const details::ParamsAndGrads& params_grads) const {
+    std::unordered_set<details::OpHandleBase*> visit;
+    for (auto op : ops) {
+      int order = 0;
+      if (visit.find(op) != visit.end()) {
+        VLOG(10) << "visited all_opt_handles:" << op->DebugString();
+        continue;
+      } else {
+        VLOG(10) << "visiting all_opt_handles:" << op->DebugString();
+      }
```
Comment (chengduoZH) [inline paddle/fluid/framework/ir/multi_devices_graph_pass/backward_optimizer_op_deps_pass.cc:102]:
```
      if (visit.find(op) != visit.end()) {
        continue;
      } 
        VLOG(10) << "visiting all_opt_handles:" << op->DebugString();
```

File: paddle/fluid/framework/ir/multi_devices_graph_pass/backward_optimizer_op_deps_pass.cc [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+  void VisitChildrens(details::OpHandleBase* op,
+                      std::unordered_set<details::OpHandleBase*>* visit,
+                      int* order) const {
+    for (auto out : op->Outputs()) {
+      for (auto* pending_op : out->PendingOps()) {
+        if (visit->find(pending_op) != visit->end()) {
+          VLOG(10) << "order:" << *order
+                   << " visited:" << pending_op->DebugString();
+          continue;
+        }
```
Comment (chengduoZH) [inline paddle/fluid/framework/ir/multi_devices_graph_pass/backward_optimizer_op_deps_pass.cc:162]:
```
        if (visit->find(pending_op) != visit->end()) {
          continue;
        }
```

File: paddle/fluid/framework/ir/multi_devices_graph_pass/backward_optimizer_op_deps_pass.cc (hunk 10 above)
Comment (gongweibao) [reply paddle/fluid/framework/ir/multi_devices_graph_pass/backward_optimizer_op_deps_pass.cc:102]: Done.

File: paddle/fluid/framework/ir/multi_devices_graph_pass/backward_optimizer_op_deps_pass.cc (hunk 11 above)
Comment (gongweibao) [reply paddle/fluid/framework/ir/multi_devices_graph_pass/backward_optimizer_op_deps_pass.cc:162]: Done.

**Review by chengduoZH — APPROVED** [review]
The code should be polished.

**Pull request merged**